File: 2D-dispersion_alex/dispersion.m
function [wv,fr,ev] = dispersion(const,wavevectors)
    
    fr = zeros(size(wavevectors,2),const.N_eig);
    if const.isSaveEigenvectors
        ev = zeros(((const.N_ele*const.N_pix)^2)*2,size(wavevectors,2),const.N_eig);
    else
        ev = [];
    end
    if const.isUseImprovement
        [K,M] = get_system_matrices_VEC(const);
    else
        [K,M] = get_system_matrices(const);
    end
    if const.isUseParallel
        parforArg = Inf;
    else
        parforArg = 0;
    end
%     for k_idx = 1:size(wavevectors,1) % USE THIS TO DEBUG
    parfor (k_idx = 1:size(wavevectors,1), parforArg)
        wavevector = wavevectors(k_idx,:);
        T = get_transformation_matrix(wavevector,const);
        Kr = T'*K*T;
        Mr = T'*M*T;
        
        if ~const.isUseGPU
            % DONT USE THE GPU WITH EIGS           
            [eig_vecs,eig_vals] = eigs(Kr,Mr,const.N_eig,const.sigma_eig);
            [eig_vals,idxs] = sort(diag(eig_vals));
            eig_vecs = eig_vecs(:,idxs);
            
%             ev(:,k_idx,:) = (eig_vecs./(diag(eig_vecs'*Mr*eig_vecs)'))'; % normalize by mass matrix
            if const.isSaveEigenvectors
                ev(:,k_idx,:) = (eig_vecs./vecnorm(eig_vecs,2,1)).*exp(-1i*angle(eig_vecs(1,:))); % normalize by p-norm, align complex angle
            end
            %             ev(:,k_idx,:) = (eig_vecs./max(eig_vecs))'; % normalize by max
%             ev(:,k_idx,:) = eig_vecs'; % don't normalize
            
            fr(k_idx,:) = sqrt(real(eig_vals));
            fr(k_idx,:) = fr(k_idx,:)/(2*pi);
        elseif const.isUseGPU
            % USE THE GPU WITH EIG
            error('GPU use is not currently developed');
            MinvK_gpu = gpuArray(full(inv(Mr)*Kr));
            [eig_vecs,eig_vals] = eig(MinvK_gpu);
            [eig_vals,idxs] = sort(diag(eig_vals));
            eig_vecs = eig_vecs(:,idxs);
            fr(k_idx,:) = sqrt(real(eig_vals(1:const.N_eig)));
            fr(k_idx,:) = fr(k_idx,:)/(2*pi);
        end
    end
    wv = wavevectors;
end

File: dispersion.py
import numpy as np
from scipy.sparse.linalg import eigs
from numpy.linalg import norm

def dispersion(wavevectors, const):
    """
    Compute the dispersion relation (frequency vs wavevector) 
    for a periodic medium using finite element matrices.
    
    Parameters:
    - wavevectors: ndarray of shape (N_k, 2), the wavevectors in the IBZ
    - const: dict with constants and flags (see details below)

    Returns:
    - wv: same as wavevectors
    - fr: frequencies of shape (N_k, N_eig)
    - ev: eigenvectors of shape (N_dof, N_k, N_eig) if saved, else None
    """
    N_k = wavevectors.shape[0]
    N_eig = const['N_eig']
    N_pix = const['N_pix']
    N_ele = const['N_ele']
    is_save_ev = const['isSaveEigenvectors']

    # Preallocate output arrays
    fr = np.zeros((N_k, N_eig))
    if is_save_ev:
        N_dof = (N_ele * N_pix)**2 * 2  # total number of degrees of freedom
        ev = np.zeros((N_dof, N_k, N_eig), dtype=np.complex128)
    else:
        ev = None

    # Get global system matrices (K and M)
    if const['isUseImprovement']:
        K, M = get_system_matrices_VEC(const)
    else:
        K, M = get_system_matrices(const)

    # Loop over all wavevectors
    for k_idx in range(N_k):
        kvec = wavevectors[k_idx, :]
        
        # Get Bloch transformation matrix for current wavevector
        T = get_transformation_matrix(kvec, const)

        # Apply the Bloch-periodic transformation
        Kr = T.conj().T @ K @ T
        Mr = T.conj().T @ M @ T

        # Solve the generalized eigenvalue problem Kr u = λ Mr u
        eig_vals, eig_vecs = eigs(Kr, M=Mr, k=N_eig, sigma=const['sigma_eig'])

        # Sort eigenvalues and eigenvectors by ascending eigenvalue
        idx = np.argsort(np.real(eig_vals))
        eig_vals = eig_vals[idx]
        eig_vecs = eig_vecs[:, idx]

        # Normalize and align eigenvectors (optional)
        if is_save_ev:
            for j in range(N_eig):
                v = eig_vecs[:, j]
                v = v / norm(v)  # normalize by L2 norm
                v = v * np.exp(-1j * np.angle(v[0]))  # align complex phase
                ev[:, k_idx, j] = v

        # Convert eigenvalues to real frequencies (Hz)
        fr[k_idx, :] = np.sqrt(np.real(eig_vals)) / (2 * np.pi)

    return wavevectors, fr, ev
